[Bug] Tracking flaky macOS CI: `dlopen … No such file` during `import libtiledbsoma`

## Describe the bug
The "macos TILEDB_EXISTS: yes TILEDBSOMA_EXISTS: no" job sometimes fails, in the "Confirm linking to installed shared objects" step ([python-so-copying.yml#L232](https://github.com/single-cell-data/TileDB-SOMA/blob/0774b69f2cba5d8aa459d828f62cd9fe0d66c685/.github/workflows/python-so-copying.yml#L232)), with error:

```
OSError: dlopen(/Users/runner/work/TileDB-SOMA/TileDB-SOMA/venv-soma/lib/python3.11/site-packages/tiledbsoma/libtiledb.dylib, 0x000A): tried: '/Users/runner/work/TileDB-SOMA/TileDB-SOMA/venv-soma/lib/python3.11/site-packages/tiledbsoma/libtiledb.dylib' (no such file)
```

<details><summary>Example full stack trace</summary>

```
Traceback (most recent call last):
  File "/Users/runner/work/TileDB-SOMA/TileDB-SOMA/venv-soma/lib/python3.11/site-packages/tiledbsoma/__init__.py", line 104, in <module>
    from . import pytiledbsoma as clib
ImportError: dlopen(/Users/runner/work/TileDB-SOMA/TileDB-SOMA/venv-soma/lib/python3.11/site-packages/tiledbsoma/pytiledbsoma.cpython-311-darwin.so, 0x0002): Library not loaded: '@rpath/libtiledbsoma.dylib'
  Referenced from: '/Users/runner/work/TileDB-SOMA/TileDB-SOMA/venv-soma/lib/python3.11/site-packages/tiledbsoma/pytiledbsoma.cpython-311-darwin.so'
  Reason: tried: '/Users/runner/work/TileDB-SOMA/TileDB-SOMA/dist/lib/libtiledbsoma.dylib' (no such file), '/Users/runner/work/TileDB-SOMA/TileDB-SOMA/external/lib/libtiledbsoma.dylib' (no such file), '/Users/runner/work/TileDB-SOMA/TileDB-SOMA/dist/lib/libtiledbsoma.dylib' (no such file), '/Users/runner/work/TileDB-SOMA/TileDB-SOMA/external/lib/libtiledbsoma.dylib' (no such file), '/usr/lib/libtiledbsoma.dylib' (no such file)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/runner/work/TileDB-SOMA/TileDB-SOMA/venv-soma/lib/python3.11/site-packages/tiledbsoma/__init__.py", line 118, in <module>
    ctypes.CDLL(os.path.join(lib_dir, libtiledb_name), mode=ctypes.RTLD_GLOBAL)
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/ctypes/__init__.py", line 376, in __init__
    self._handle = _dlopen(self._name, mode)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^
OSError: dlopen(/Users/runner/work/TileDB-SOMA/TileDB-SOMA/venv-soma/lib/python3.11/site-packages/tiledbsoma/libtiledb.dylib, 0x000A): tried: '/Users/runner/work/TileDB-SOMA/TileDB-SOMA/venv-soma/lib/python3.11/site-packages/tiledbsoma/libtiledb.dylib' (no such file)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/Users/runner/work/TileDB-SOMA/TileDB-SOMA/venv-soma/lib/python3.11/site-packages/tiledbsoma/__init__.py", line 121, in <module>
    ctypes.CDLL(libtiledb_name, mode=ctypes.RTLD_GLOBAL)
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/ctypes/__init__.py", line [37](https://github.com/single-cell-data/TileDB-SOMA/actions/runs/8191829188/job/22401948329?pr=2235#step:16:38)6, in __init__
    self._handle = _dlopen(self._name, mode)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^
OSError: dlopen(libtiledb.dylib, 0x000A): tried: 'libtiledb.dylib' (no such file), '/usr/lib/libtiledb.dylib' (no such file), '/Users/runner/work/TileDB-SOMA/TileDB-SOMA/libtiledb.dylib' (no such file)
Error: Process completed with exit code 1.
```
</details> 

Running list of instances of this failure that I've noticed:
- [3/6/24](https://github.com/single-cell-data/TileDB-SOMA/actions/runs/8180076623/job/22367393933?pr=2227#step:16:37) (@johnkerl)
- [3/7/24](https://github.com/single-cell-data/TileDB-SOMA/actions/runs/8182105341/job/22372928875#step:16:37) (me)
- [3/7/24](https://github.com/single-cell-data/TileDB-SOMA/actions/runs/8191829188/job/22401948329?pr=2235#step:16:37) (me)
- [3/15/24](https://github.com/single-cell-data/TileDB-SOMA/actions/runs/8300380263/job/22718191691?pr=1960#step:16:37) (me)
- [3/21/24](https://github.com/single-cell-data/TileDB-SOMA/actions/runs/8383229307/job/22958534492?pr=2256#step:16:48) (me)

In all cases, 1 re-run was sufficient to mine a successful run.

### Aside: mitigation + possible separate segfault issue
https://github.com/single-cell-data/TileDB-SOMA/pull/2229 should make it less common (by running the job in question less frequently/superfluously).

Note that a similar flaky CI job was observed there, by @jdblischak ([GHA link](https://github.com/single-cell-data/TileDB-SOMA/actions/runs/8180364638/job/22368243175#step:4:78)), in job "macos TILEDB_EXISTS: yes TILEDBSOMA_EXISTS: yes", step [Build and install libtiledbsoma](https://github.com/single-cell-data/TileDB-SOMA/blob/92593fd72acb58ebb283f7e4391a7d2b426b92e4/.github/workflows/python-so-copying.yml#L155-L168):

```
clang: error: unable to execute command: Segmentation fault: 11
```

<details><summary>Full stack trace</summary>

```
-- Install prefix is /Users/runner/work/TileDB-SOMA/TileDB-SOMA/external.
-- The C compiler identification is AppleClang 14.0.0.14000029
-- The CXX compiler identification is AppleClang 14.0.0.14000029
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /Applications/Xcode_14.2.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /Applications/Xcode_14.2.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Starting TileDB-SOMA superbuild.
-- Found TileDB: /Users/runner/work/TileDB-SOMA/TileDB-SOMA/external/lib/libtiledb.dylib
-- Could NOT find spdlog (missing: spdlog_DIR)
-- Adding spdlog as an external project
-- Not found clang-tidy
-- Not found clang-format
-- Configuring done (4.4s)
-- Generating done (0.0s)
-- Build files have been written to: /Users/runner/work/TileDB-SOMA/TileDB-SOMA/build-libtiledbsoma
/Users/runner/work/_temp/f7f53cdf-a50d-4785-9ee7-cb72edffd8d5.sh: line 8: nproc: command not found
[  6%] Creating directories for 'ep_spdlog'
[ 12%] Performing download step (download, verify and extract) for 'ep_spdlog'
-- ep_spdlog download command succeeded.  See also /Users/runner/work/TileDB-SOMA/TileDB-SOMA/build-libtiledbsoma/externals/src/ep_spdlog-stamp/ep_spdlog-download-*.log
[ 18%] No update step for 'ep_spdlog'
[ 25%] Performing patch step for 'ep_spdlog'
patching file include/spdlog/logger-inl.h
[ 31%] Performing configure step for 'ep_spdlog'
-- ep_spdlog configure command succeeded.  See also /Users/runner/work/TileDB-SOMA/TileDB-SOMA/build-libtiledbsoma/externals/src/ep_spdlog-stamp/ep_spdlog-configure-*.log
[ 37%] Performing build step for 'ep_spdlog'
CMake Error at /Users/runner/work/TileDB-SOMA/TileDB-SOMA/build-libtiledbsoma/externals/src/ep_spdlog-stamp/ep_spdlog-build-Release.cmake:37 (message):
-- stdout output is:
  Command failed: 2
[ 10%] Building CXX object CMakeFiles/spdlog.dir/src/spdlog.cpp.o

[ [20](https://github.com/single-cell-data/TileDB-SOMA/actions/runs/8180364638/job/22368243175#step:4:21)%] Building CXX object CMakeFiles/spdlog.dir/src/stdout_sinks.cpp.o
   '/Applications/Xcode_14.2.app/Contents/Developer/usr/bin/make'
[ 30%] Building CXX object CMakeFiles/spdlog.dir/src/color_sinks.cpp.o

  See also

    /Users/runner/work/TileDB-SOMA/TileDB-SOMA/build-libtiledbsoma/externals/src/ep_spdlog-stamp/ep_spdlog-build-*.log


CMake Error at /Users/runner/work/TileDB-SOMA/TileDB-SOMA/build-libtiledbsoma/externals/src/ep_spdlog-stamp/ep_spdlog-build-Release.cmake:47 (message):
  Stopping after outputting logs.


make[2]: *** [externals/src/ep_spdlog-stamp/ep_spdlog-build] Error 1
make[1]: *** [CMakeFiles/ep_spdlog.dir/all] Error 2
make: *** [all] Error 2
[ 40%] Building CXX object CMakeFiles/spdlog.dir/src/file_sinks.cpp.o
[ 50%] Building CXX object CMakeFiles/spdlog.dir/src/async.cpp.o
[ 60%] Building CXX object CMakeFiles/spdlog.dir/src/cfg.cpp.o
[ 70%] Building CXX object CMakeFiles/spdlog.dir/src/fmt.cpp.o

-- stderr output is:
clang: error: unable to execute command: Segmentation fault: 11
clang: error: clang frontend command failed due to signal (use -v to see invocation)
Apple clang version 14.0.0 (clang-1400.0.29.202)
Target: x86_64-apple-darwin[21](https://github.com/single-cell-data/TileDB-SOMA/actions/runs/8180364638/job/22368243175#step:4:22).6.0
Thread model: posix
InstalledDir: /Applications/Xcode_14.2.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
clang: note: diagnostic msg: 
********************

PLEASE ATTACH THE FOLLOWING FILES TO THE BUG REPORT:
Preprocessed source(s) and associated run script(s) are located at:
clang: note: diagnostic msg: /var/folders/h1/8hndypj13nsbj5pn4xsnv1tm0000gn/T/spdlog-510f3d.cpp
clang: note: diagnostic msg: /var/folders/h1/8hndypj13nsbj5pn4xsnv1tm0000gn/T/spdlog-510f3d.sh
clang: note: diagnostic msg: Crash backtrace is located in
clang: note: diagnostic msg: /Users/runner/Library/Logs/DiagnosticReports/clang_<YYYY-MM-DD-HHMMSS>_<hostname>.crash
clang: note: diagnostic msg: (choose the .crash file that corresponds to your crash)
clang: note: diagnostic msg: 

********************
make[5]: *** [CMakeFiles/spdlog.dir/src/spdlog.cpp.o] Error [25](https://github.com/single-cell-data/TileDB-SOMA/actions/runs/8180364638/job/22368243175#step:4:26)4
make[5]: *** Waiting for unfinished jobs....
make[4]: *** [CMakeFiles/spdlog.dir/all] Error 2
make[3]: *** [all] Error 2

Error: Process completed with exit code 2.
```
</details>

Can file a separate issue for that, if we keep seeing it.

--- a/.github/workflows/python-so-copying.yml
+++ b/.github/workflows/python-so-copying.yml
@@ -139,7 +139,7 @@ jobs:
           ./venv-soma/bin/python -c "import tiledbsoma; print(tiledbsoma.pytiledbsoma.version())"
 
   macos:
-    runs-on: macos-12
+    runs-on: macos-13
     name: "macos TILEDB_EXISTS: ${{ matrix.TILEDB_EXISTS }} TILEDBSOMA_EXISTS: ${{ matrix.TILEDBSOMA_EXISTS }}"
     strategy:
       fail-fast: false
@@ -153,6 +153,8 @@ jobs:
       - uses: actions/checkout@v4
         with:
           fetch-depth: 0 # for setuptools-scm
+      - name: Check if System Integrity Protection (SIP) is enabled
+        run: csrutil status
       - name: Install pre-built libtiledb
         if: ${{ matrix.TILEDB_EXISTS == 'yes' }}
         run: |

